docs/rfcs/logging-before-config-resolution.md

```diff
@@ -0,0 +1,62 @@
+# How To Log Before Config Resolution
+
+## Overview
+
+The OpenTelemetry Collector supports configuring a primary logger that the collector and its components use to write logs.
+This logger cannot be created until the user's configuration has been completely resolved.
+There is a need to write logs during the collector start-up, before the primary logger is instantiated, such as during
+configuration resolution or config validation. This document describes
+
+- why providing logging capabilities during startup is important
+- the current (as of v0.99.0) behavior of the Collector
+- different solutions to the problem
+- the accepted solution
+
+## Why Logging During Startup is Important
+
+When the collector is starting it tries to resolve user configuration as quickly as possible.
+But the Collector's configuration resolution strategy is not trivial - it allows for complex interactions between
+multiple, different config sources that must all resolve without error. During this process important information could
+be shared with users such as:
+- [Warnings about deprecated syntax](https://github.com/open-telemetry/opentelemetry-collector/issues/9162)
+- [Warnings about undesired, but handled, situations](https://github.com/open-telemetry/opentelemetry-collector/issues/5615)
+- Debug information
+
+## Requirements for any solution
+
+1. Once the primary logger is instantiated, it should be usable anywhere in the Collector that does logging.
+
+## Current behavior
+
+As of v0.99.0, the collector does not provide a way to log before the primary logger is instantiated.
+
+## Solutions
+
+### Buffer Logs in Memory and Write Them Once the Primary Logger Exists
+
+The Collector could provide a temporary logger that, when written to, keeps the logs in memory. These logs could then
+be passed to the primary logger to be written out in the properly configured format/level.
+
+Benefits:
+- Logs are written in the user-specified format/level
+
+Downsides:
+- If an error occurs before the primary logger is created the buffered logs will not be written out
+- If the primary logger is used to write any logs before the buffered logs are passed, logs will be out of order
```

Comment:
I think we can fix this if we can find a good solution. This is an implementation detail of the Collector<->service interface.

Reply:
I agree that we can prevent this downside from happening with good coding standards, but ultimately it is a possibility.

Reply:
This could be mitigate at initialization of the logger no? Once the config has been resolved, it should be possible to output the buffered logs before making the logger available to components, or am i missing something here

Reply:
Okay, I see more of what you mean now. Yes, I suppose that we can't guarantee logs will have completely correct ordering. I do think we could fairly easily achieve it so long as we find a way to have the service itself print out the confmap logs. The only places we have a zap logger will be in the confmap components and in the service. I think it would be fairly straightforward to unmarshal the config, create the logger, then log the confmap logs before anything else is done.

I think we should change the statement to say that it opens up the opportunity for them to be printed out of order, not that they necessarily will be printed out of order.

Reply:
As long as logs are timestamped (which they always should be), this isn't much of an issue. Ultimately the timestamps define the order, rather than their order of appearance in a medium.

Reply:
> As long as logs are timestamped (which they always should be), this isn't much of an issue. Ultimately the timestamps define the order, rather than their order of appearance in a medium.

I'm ok with this, I'll add to the requirements: `log timestamps must be preserved no matter when they end up being written to stdout/stderr